File: java/leetcode_grind/Day552.java
package leetcode_grind;

public class Day552 {
    // https://leetcode.com/problems/find-the-maximum-sum-of-node-values/description
    static class Solution1 {
        public long maximumValueSum(int[] nums, int k, int[][] edges) {
            var n = nums.length;

            var dp = new long[n + 1][2];
            dp[n][1] = 0;
            dp[n][0] = -(1 << 31);

            for (var i = n - 1; i >= 0; i--) {
                for (var j = 0; j < 2; j++) {
                    var xor = (nums[i] ^ k) + dp[i + 1][j ^ 1];
                    dp[i][j] = Math.max(xor, nums[i] + dp[i + 1][j]);
                }
            }

            return dp[0][1];
        }
    }
}

File: src/leetcode_grind/day_552.rs
// https://leetcode.com/problems/find-the-maximum-sum-of-node-values/description
pub fn maximum_value_sum(nums: Vec<i32>, k: i32, edges: Vec<Vec<i32>>) -> i64 {
    let mut sum = 0i64;
    let mut count = 0;

    let mut positive_min = (1 << 30);
    let mut negative_max = -(1 << 30);

    for &node_val in &nums {
        let mut operated_node_value = node_val ^ k;
        sum += node_val as i64;
        let mut net_change = operated_node_value - node_val;

        if net_change > 0 {
            positive_min = positive_min.min(net_change);
            sum += net_change as i64;
            count += 1;
        } else {
            negative_max = negative_max.max(net_change);
        }
    }

    if count % 2 == 0 {
        return sum;
    }

    return (sum - positive_min as i64).max(sum + negative_max as i64);
}
